src/include/plan_node/abstract_plan_node.h

```diff
@@ -0,0 +1,141 @@
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "common/hash_util.h"
+#include "common/json.h"
+#include "plan_node/output_schema.h"
+#include "plan_node/plan_node_defs.h"
+
+// TODO(Gus,Wen): Add equaility operator and hash function support for output_schema
+
+namespace terrier::plan_node {
+//===--------------------------------------------------------------------===//
+// Abstract Plan Node
+//===--------------------------------------------------------------------===//
+
+class AbstractPlanNode {
+ public:
+  explicit AbstractPlanNode(std::shared_ptr<OutputSchema> output_schema) : output_schema_(std::move(output_schema)) {}
+
+  // For Deserialization and DDL statements
+  AbstractPlanNode() = default;
+
+  virtual ~AbstractPlanNode() = default;
+
+  //===--------------------------------------------------------------------===//
+  // Children Helpers
+  //===--------------------------------------------------------------------===//
+
+  void AddChild(std::unique_ptr<AbstractPlanNode> &&child) { children_.emplace_back(std::move(child)); }
+
+  const std::vector<std::unique_ptr<AbstractPlanNode>> &GetChildren() const { return children_; }
+
+  size_t GetChildrenSize() const { return children_.size(); }
+
+  const AbstractPlanNode *GetChild(uint32_t child_index) const {
+    TERRIER_ASSERT(child_index < children_.size(),
+                   "index into children of plan node should be less than number of children");
+    return children_[child_index].get();
+  }
+
+  //===--------------------------------------------------------------------===//
+  // Accessors
+  //===--------------------------------------------------------------------===//
+
+  // Each sub-class will have to implement this function to return their type
+  // This is better than having to store redundant types in all the objects
+  virtual PlanNodeType GetPlanNodeType() const { return PlanNodeType::ABSTRACTPLAN; }
+
+  // Get a string representation for debugging
+  virtual const std::string GetInfo() const;
+
+  // Get the output schema for the plan node. The output schema contains information on columns of the output of
+  // the plan node operator
+  std::shared_ptr<OutputSchema> GetOutputSchema() const { return output_schema_; }
+
+  // Get the estimated cardinality of this plan
+  int GetEstimatedCardinality() const { return estimated_cardinality_; }
+
+  // FOR TESTING ONLY. This function should only be called during construction of plan (ConvertOpExpression) or
+  // for tests.
+  void SetEstimatedCardinality(int cardinality) { estimated_cardinality_ = cardinality; }
```

Comment:
It is a good idea to reflect it in the method name if it's test only. e.g. `SetEstimatedCardinalityForUnitTest`

We can talk about general best practices for this kind of things in the group meeting.

Reply:
I removed the method as we're using a builder pattern now.